feat: internal audit logging system

oliverqx opened the pull request:
/claim #1461
## What does this PR do?

This PR adds an internal audit logging system to cal.com. Its designed to log actions using the expected audit-log clients. 

<!-- Please include a summary of the change and which issue is fixed. Please also include relevant motivation and context. List any dependencies that are required for this change. -->

- Fixes #1461  (GitHub issue number)
- Fixes CAL-1710 (Linear issue number - should be visible at the bottom of the GitHub issue description)

I recorded a 10 minute video showing my work. In the video I show how this internal audit logging system enables audit log app development by developing the BoxyHQ integration. You can find the video here: https://youtu.be/Nbi8Tr1Ny9E

## Mandatory Tasks (DO NOT REMOVE)

- [ ] I have self-reviewed the code (A decent size PR without self-review might be rejected)
- [ ] I have added a Docs issue [here](https://github.com/calcom/docs/issues/new) if this PR makes changes that would require a [documentation change](https://docs.cal.com)
- [ ] I have added or modified automated tests that prove my fix is effective or that my feature works (PRs might be rejected if logical changes are not properly tested)

## How should this be tested?

<!-- Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration. Write details that help to start the tests -->

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

oliverqx committed:
chore: update App table

add settings row.

oliverqx commented:
I'm trying to compile a list of all possible actions that can be logged by the audit system. So far I've this:

```
Webhooks
	- Create
	- Modify
		Event Triggers
		Subscriber URL
		Secret
		Payload Template
	- Enable	
	- Disable
	- Delete
API Keys
	- Create
	- Modify
		- Name
Schedule
	- Modified
		Available Hours
		Date Overrides
		Timezone
		Metadata
	- Created
	- Deleted
Bookings
	- Modified
		Reschedule
		Edit Location
	- Created
	- Cancelled
	- Reschedule Requested
Event-Types
	- Modified
		Event Setup
		Availability
		Limits
		Advanced
		Recurring
		Apps
		Workflows
		Webhooks
		Visibility
	- Created
	- Duplicate
	- Delete
Teams
	- Modified
		Members
		Profile
		Apperance
		Availability
	- Created
	- Duplicate
	- Deleted
User Profile
	- Modified
		Profile
		General
		Calendar	
		Conferencing
		Appearance
		Out of Office
	- Created
	- Delete
Password
	- Modified
	- Created
	- Delete
Impersonation
	- On
	- Off
```

oliverqx commented:
I'm trying to compile a list of all possible actions that can be logged by the audit system. So far I've this:

```
Webhooks
	- Create
	- Modify
		Event Triggers
		Subscriber URL
		Secret
		Payload Template
	- Enable	
	- Disable
	- Delete
API Keys
	- Create
	- Modify
		- Name
Schedule
	- Modified
		Available Hours
		Date Overrides
		Timezone
		Metadata
	- Created
	- Deleted
Bookings
	- Modified
		Reschedule
		Edit Location
	- Created
	- Cancelled
	- Reschedule Requested
Event-Types
	- Modified
		Event Setup
		Availability
		Limits
		Advanced
		Recurring
		Apps
		Workflows
		Webhooks
		Visibility
	- Created
	- Duplicate
	- Delete
Teams
	- Modified
		Members
		Profile
		Apperance
		Availability
	- Created
	- Duplicate
	- Deleted
User Profile
	- Modified
		Profile
		General
		Calendar	
		Conferencing
		Appearance
		Out of Office
	- Created
	- Delete
Password
	- Modified
	- Created
	- Delete
Impersonation
	- On
	- Off
```

cc: @PeerRich

oliverqx commented:
cc: @baileypumfleet

oliverqx committed:
feat: create audit logs manager dummy and utils

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/calcom/cal.com?pullRequest=14950) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/calcom/cal.com?pullRequest=14950) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/calcom/cal.com?pullRequest=14950) it.</sub>

oliverqx committed:
fix: temporarily hard coded appKeys

oliverqx committed:
fix: add parameter to handleAuditLogTrigger

oliverqx committed:
feat: add audit log when new booking is created

oliverqx committed:
feat: add audit log to all webhook actions

this commit adds an audit log trigger whenever webhooks are triggered.

oliverqx committed:
fix: import statements typo

oliverqx committed:
revert: yarn.lock

oliverqx committed:
feat(schema): extend 'AppCategories' enum with new value "auditLogs"

oliverqx committed:
feat(utils): add auditLogs to AppCategory utils

oliverqx committed:
add(icon): add auditLogs icon to emptyIcon object

oliverqx committed:
add: audit logs locale

commit includes both english and spanish

oliverqx committed:
feat: add audit logs integration app template

oliverqx committed:
fix: modify template slug

oliverqx committed:
fix: typo in category name

oliverqx committed:
add: audit logs app category image

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
chore: improve types in app template

oliverqx committed:
fix: update key schema to match expected by implementation

oliverqx committed:
refactor: improve types

oliverqx committed:
chore: add AuditLogEvent type

oliverqx committed:
fix: implement AuditLogEvent in template audit log manager

oliverqx committed:
remove: unnecessary comments

oliverqx committed:
chore: add example code

oliverqx committed:
feat: implement audit log event in trigger

oliverqx committed:
chore: add types

oliverqx committed:
chore: improve types by implementing auditLog enums

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
chore: update schema and add constants

constants will be used to render audit logs app settings.

oliverqx committed:
feature: audit log booking events translations

oliverqx committed:
fix: filename typo

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
feat: add trpc endpoint addAppCredentialById

Will be used by AppSettingsInterface to retrieve active credential information

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
fix: typing

oliverqx committed:
feat: add settings panel

Panel will let users update credentials, show system status, and turn on/off logs. It can be easily extended to do much more if a future integration/app requires it.

oliverqx committed:
fix: typos and missing schema entry

oliverqx committed:
feat: add ping endpoint

will be able to tell if system is operational

oliverqx committed:
feat: implement settings panel

oliverqx committed:
fix: interface AuditLogsManager from report to reportEvent

oliverqx committed:
fix: remove app settings

not implemented yet, changes were committed by mistake

oliverqx committed:
feat: improve readability/dx

oliverqx committed:
fix: import statements

oliverqx committed:
fix: improve typing of key object in credentials

oliverqx committed:
feat: UI/UX touchup for template audit log implementation app

oliverqx committed:
feat: ping endpoint and ui implemented

oliverqx committed:
feat: audit log event toggles UX/UI

oliverqx committed:
refactor: move locale location

oliverqx committed:
fix: return settings when requesting credential

oliverqx committed:
feat: add update credential settings to trpc router

oliverqx committed:
feat: add settings to Credential schema

oliverqx committed:
feat: implement audit log event toggles logic

oliverqx committed:
fix: update imports

oliverqx committed:
feat: implement event settings to audit log system

event will not be sent/logged if its been disabled by the administrator

oliverqx committed:
feat: remove hard coded params

oliverqx committed:
fix: implement handleAuditLogTrigger changes to every call

oliverqx committed:
fix: update description and name for audit-log-implementation template

oliverqx committed:
chore: fix types

oliverqx committed:
fix: export all definitions from template

oliverqx committed:
feat: make getAuditLogManager generic

removed hard coded options, will return manager based on appId

oliverqx committed:
refactor: improve dx

oliverqx committed:
fix: modify export name for auditLogManager

oliverqx committed:
fix: properly export utilities from audit log template

oliverqx committed:
fix: update utils imports

oliverqx committed:
fix: typos and remove commented code

oliverqx committed:
fix: modify constants

oliverqx committed:
feat: improve app setting options types

- Better DX
- Easy to extend options in a typesafe way
- Enables translation of settings navigation panel

oliverqx committed:
fix: add property to app settings options

- Vertical Tabs does not take in a translation object, its isolated. Translations would have to go in common.json. Feels rather dirty that way.

oliverqx committed:
feat: add loading state to credentials form

oliverqx committed:
fix: spacing

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx commented:
Hello,

Just wanted to let you guys know that I think my PR is ready for review... I do need some feedback in one small part: I still don't have a list of events to be audited/reported. I did compile one but would like to hear you guys' [thoughts on it.](https://github.com/calcom/cal.com/pull/14950#issuecomment-2103199128)

I also just recorded [a video](https://youtu.be/Nbi8Tr1Ny9E) where I show how my solution makes it easy to create custom audit log integrations by live coding the BoxyHQ Retraced app in less than 10 minutes.

Would really appreciate the feedback! 🚀 

cc:
@baileypumfleet 
@PeerRich 
@zomars

oliverqx committed:
fix: update type name

oliverqx committed:
chore: remove console.logs

oliverqx committed:
feat: create trpc middleware plugin

oliverqx committed:
fix: return genericManager as default

was previously returning function without initiating default manager

oliverqx committed:
chore: update types

oliverqx committed:
feat: create trpc procedure with auditLogMiddleware

V1, will run whenever appCredentials are updated

oliverqx committed:
Merge branch 'feature.logging' of https://github.com/oliverqx/cal.com into feature.logging

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx commented:
cc: 
@keithwillcode

zomars reviewed (changes requested):
Thank you for your contribution 🙏 Will leave some comments before starting an official review.

Self- review is missing. Please go into the [Files section](https://github.com/calcom/cal.com/pull/14950/files) and leave comments on "why" things are done certain ways. 

For example, please comment on the `schema.prisma` file. Why do you feel those changes are needed. Why on those models? What alternatives did you consider and why did you end up choosing this one.

Whatever piece of info that helps reviewers into understanding your mind context will allow us to get feedback sooner for you.

oliverqx commented:
> Thank you for your contribution 🙏 Will leave some comments before starting an official review.
> 
> Self- review is missing. Please go into the [Files section](https://github.com/calcom/cal.com/pull/14950/files) and leave comments on "why" things are done certain ways.
> 
> For example, please comment on the `schema.prisma` file. Why do you feel those changes are needed. Why on those models? What alternatives did you consider and why did you end up choosing this one.
> 
> Whatever piece of info that helps reviewers into understanding your mind context will allow us to get feedback sooner for you.

will do asap thanks!

oliverqx committed:
fix: missing types and typos

oliverqx committed:
chore: update AuditLogEvent type

Will be used by audit log trigger

oliverqx committed:
chore: add request-ip dependency to @calcom/web

oliverqx committed:
chore: temporarily comment out handleAuditLogTrigger

oliverqx committed:
feat: update handleAuditLogTrigger

- updates handleAuditLogTrigger
- implements booking creation log (hard coded)

oliverqx committed:
feat: create bookings zod schema

- Creates zod schema
- Implements schema to parse event

oliverqx committed:
feat(prisma): add BOOKING_CONFIRMED as trigger event to AuditLogBookingTriggerEvents schema

oliverqx committed:
feat: add auditLogTrigger on cancel and rejection

oliverqx commented on packages/prisma/schema.prisma:
The architecture of the audit logs feature will enable users to decide which events are reported through a nice UI. This is part of the default internal behaviour, and even though third party implementations can disable it if they choose to, I still needed a way to store disabled events. 

Settings is used to store an object of type: { disabledEvents: AuditLogTriggerEvents[] }, handleAuditLogTrigger will loop through all credentials and report the event if event is not found in disableEvents. 

Having settings stored directly in the app’s Credential allows handleAuditLogTrigger to get all necessary information in a single query and it also allows for isolation of preferences between app installations.

For example, BoxyHQ Retraced could be installed several times per user/team/organization and each installation could have its own stream/project to report to, and unique preferences as to what to report and what not to report. 

I considered using App.settings as the storage but this would define settings globally to all BoxyHQ Retraced instances, disabling isolation.

oliverqx committed:
fix: remove settings from App

settings will be stored at Credential level to enable settings isolation between installations

oliverqx committed:
fix: revert git ignore changes

oliverqx commented on packages/trpc/server/routers/publicViewer/i18n.handler.ts:
Considering the amount of text that would be added to "common", I though having separate files for translations would be much better

oliverqx commented on packages/prisma/schema.prisma:
```diff
@@ -171,6 +171,8 @@ model Credential {
   invalid              Boolean?              @default(false)
   CalendarCache        CalendarCache[]
+  /// Note: settings type depends on implementation/slug this credential belongs to.
+  settings             Json                  @default("{ \"empty\": \"true\" }")
```
The architecture of the audit logs feature will enable users to decide which events are reported through a nice UI. This is part of the default internal behaviour and disabled events are stored in Credential.settings. 

Settings is used to store an object of type: { disabledEvents: AuditLogTriggerEvents[] }, handleAuditLogTrigger will loop through all credentials and report the event if event is not found in disableEvents. 

Having settings stored directly in the app’s Credential allows for isolation of preferences between app installations.

For example, BoxyHQ Retraced could be installed several times per user/team/organization and each installation could have its own stream/project to report to, and unique preferences as to what to report and what not to report. 

I considered using App.settings as the storage, but this would define settings globally to all BoxyHQ Retraced instances, disabling isolation.

oliverqx commented on packages/prisma/schema.prisma:
```diff
@@ -700,6 +702,33 @@ enum PaymentOption {
+  WORKFLOWS
+  SETTINGS
+}
+
```
Following Webhook feature design, I created these enums. They are used throughout the feature for typesafety, to render the eventToggles UI (where users can disable/enable event reporting) and for UI translations.

oliverqx commented on packages/prisma/schema.prisma:
```diff
@@ -815,6 +844,7 @@ enum AppCategories {
   // Wherever video is in use, conferencing should also be used for legacy apps can have it.
   conferencing
   crm
+  auditLogs
```
Creates a new App Category, which will house all auditLog implementations

oliverqx commented on apps/web/components/apps/AppList.tsx:
```diff
@@ -119,7 +119,10 @@ export const AppList = ({ data, handleDisconnect, variant, listClassName }: AppL
+        <AppSettings
+          slug={item.slug}
+          credentialId={item.credentialOwner?.credentialId || item.userCredentialIds[0]}
+        />
```
I need credentialId available in AppSettings so that credential information can be seen and updated in the credentials form. Only way to get this information is by drilling it down from this component.

oliverqx commented on apps/web/package.json:
```diff
@@ -126,6 +127,7 @@
     "react-use-intercom": "1.5.1",
     "recoil": "^0.7.7",
     "remove-markdown": "^0.5.0",
+    "request-ip": "^3.3.0",
```
Used to get the actor's source ip.

oliverqx commented on apps/web/pages/apps/installed/[category].tsx:
```diff
@@ -50,6 +50,7 @@ const IntegrationsContainer = ({
     video: "video", // deprecated
     messaging: "mail",
     crm: "contact",
+    auditLogs: "notebook-pen",
```
App category icon looks like this:

oliverqx commented on apps/web/public/static/locales/en/audit-logs.json:
```diff
@@ -0,0 +1,47 @@
+{
+  "events": {
+    "BOOKING_CREATED": {
```
As you can see here, UI translations use Enum AuditLogTriggerEvent values

oliverqx commented on packages/app-store-cli/src/components/AppCreateUpdateForm.tsx:
```diff
@@ -93,6 +93,7 @@ export const AppForm = ({
         { label: "CRM", value: "crm" },
         { label: "Messaging", value: "messaging" },
         { label: "Payment", value: "payment" },
+        { label: "Audit Log", value: "auditLogs" },
```
Need category available to properly categorize apps. If auditLog implementations are not categorizes properly, they will not work

oliverqx commented on packages/app-store/_components/AppSettings.tsx:
```diff
@@ -2,7 +2,7 @@ import { AppSettingsComponentsMap } from "@calcom/app-store/apps.browser.generat
 import { DynamicComponent } from "./DynamicComponent";
 
-export const AppSettings = (props: { slug: string }) => {
+export const AppSettings = (props: { slug: string; credentialId: number }) => {
```
App settings will drill down credentialId to the dynamic component. Only audit log implementations expect this atm, but other apps wont be hurt by the change.

oliverqx commented on packages/app-store/templates/audit-log-implementation/api/ping.ts:
```diff
@@ -0,0 +1,47 @@
+  credentialId: z.number(),
+});
+
+export async function handler(req: NextApiRequest, res: NextApiResponse) {
```
This api endpoint is used by the SystemStatus component. Thought it would be nice to have some sort of visual feedback. If credentials and system is working then users will know right away. Also pinging incurs in an event being reported. This is one of the events that can not/should not be disabled at all.

oliverqx commented on packages/app-store/templates/audit-log-implementation/components/AppSettingsInterface.tsx:
```diff
@@ -0,0 +1,132 @@
+import { zodResolver } from "@hookform/resolvers/zod";
```
This component will house all settings, and features for any audit log implementation. By default it exposes a CredentialForm and the EventToggles interface. Users can easily extend this to include things like a logger-viewer ( in BoxyHQs case) Default components can also be removed

oliverqx commented on packages/app-store/templates/audit-log-implementation/components/AuditLogEventToggles.tsx:
```diff
@@ -0,0 +1,117 @@
+import { useState } from "react";
```
This component allows users to enable/disable events. This action itself will trigger a report, and this the only other event that can not/should not be disabled.

oliverqx commented on packages/app-store/templates/audit-log-implementation/components/ManagedAuditLogEventDialog.tsx:
```diff
@@ -0,0 +1,36 @@
+import { useLocale } from "@calcom/lib/hooks/useLocale";
```
Event disabling/enabling is done per event only. This Dialog confirms user desired action before executing it. Kinda like how when you want to delete a repository github makes you write down its name. This is much simpler, only a button click, but still nice UX.

oliverqx commented on packages/app-store/templates/audit-log-implementation/components/NavigationPanel.tsx:
```diff
@@ -0,0 +1,8 @@
+import { VerticalTabs } from "@calcom/ui";
```
Since the app interface can be quite complex, this component is meant to enable easy navigation between all available panels. Its also easily extendable, getAppSettingsOptions can be altered to add/delet tabs when panels are added/removed

oliverqx commented on packages/app-store/templates/audit-log-implementation/config.json:
```diff
@@ -0,0 +1,16 @@
+{
```
This is a template, will be loaded with all that is needed for a third party integration.

oliverqx commented on packages/app-store/templates/audit-log-implementation/lib/AuditLogManager.ts:
```diff
@@ -0,0 +1,28 @@
+import { getGenericAuditLogClient } from "@calcom/features/audit-logs/lib/getGenericAuditLogClient";
```
GenericAuditLogManager is just a wrapper around whatever client you give it. This is the default manager which can and should be updated to make it useful. Out of the box it will only console.log whatever event its been given.

oliverqx commented on packages/features/audit-logs/constants.ts:
```diff
@@ -0,0 +1,79 @@
+import { DefaultAppSettingsOptions } from "./types";
+import { getHref } from "./utils";
+
+export const availableTriggerEvents: { [key: string]: string | any } = {
```
availableTriggerEvents and availableTriggerTargets are used to render the Event Toggles

oliverqx commented on packages/features/audit-logs/constants.ts:
```diff
@@ -0,0 +1,79 @@
+  // },
+};
+
+export function getDefaultAppSettings(credentialId: number): DefaultAppSettingOptionEntry[] {
```
Used to render default available panels: Credentials and Triggers. This itself should never be modified, extra available panels should be added at the getAppSettingsOptions function, which is defined at the app/integration level.

oliverqx commented on packages/features/audit-logs/lib/getAuditLogManager.ts:
```diff
@@ -0,0 +1,47 @@
+import { safeStringify } from "@calcom/lib/safeStringify";
+import type { Credential } from "@calcom/prisma/client";
+
+const auditLogImplementations = {
```
Wanted to create this automatically, kinda like how all generated files are created but since so very few apps are using this i thought it would be easier to manually edit the object.  If a new integration is developed but not defined in this object, the integration's AuditLogManager will not be found and our genericImplementation will be used. (reporting events is just a console.log)

oliverqx commented on packages/features/audit-logs/lib/handleAuditLogTrigger.ts:
```diff
@@ -0,0 +1,209 @@
+  } else return clauses[0];
+};
+
+export async function handleAuditLogTrigger({
```
this is a work in progress. Comments show what the function is supposed to be in charge of. Its a bit dirty atm and will be done in a couple of days once all eventTriggers have been integrated

oliverqx commented on packages/features/audit-logs/lib/trpc-plugin.ts:
```diff
@@ -0,0 +1,25 @@
+import type { UserFromSession } from "@calcom/trpc/server/middlewares/sessionMiddleware";
```
Will be used in all trpc endpoints that perform an action that must be reported. Design is not finalized as edge cases arise it will be modified. Should be done in a couple of days

oliverqx committed:
fix: removed unused code

oliverqx commented on packages/trpc/server/routers/loggedInViewer/_router.tsx:
```diff
@@ -168,6 +170,21 @@ export const loggedInViewerRouter = router({
       return UNSTABLE_HANDLER_CACHE.appCredentialsByType({ ctx, input });
     }),
 
+  appCredentialById: authedProcedure.input(ZAppCredentialByIdInputSchema).query(async ({ ctx, input }) => {
```
Needed to fill Credential Form. Only thing left is returning parital/cloaked ApiKey to avoid exposing it, but projectID and endpoint are definitely something the user has to see.

oliverqx commented on packages/trpc/server/routers/viewer/apps/_router.tsx:
```diff
@@ -107,6 +112,26 @@ export const appsRouter = router({
       });
     }),
 
+  updateCredentialSettings: authedProcedure
```
Endpoint used to update disabledEvents

oliverqx commented on packages/trpc/server/routers/viewer/apps/updateAppCredentials.handler.ts:
```diff
@@ -21,7 +21,6 @@ export const updateAppCredentialsHandler = async ({ ctx, input }: UpdateAppCrede
   const credential = await prisma.credential.findFirst({
     where: {
       id: input.credentialId,
-      userId: user.id,
```
I need to check this. Dont really know why. Might be an error.

oliverqx reviewed (commented):
This is what I have so far. handleAuditLogTrigger and trpc audit log middleware will definitely change. I think I'm also missing some eventTriggers and evetTargets. And I'm almost done with a [boxyHQ template manager](https://boxyhq.com/docs/retraced/advanced/display-templates), which will allow users to control how events are rendered in the viewer: 

BOOKING_CANCELLED: `{{ actor.name }} cancelled booking {{ fields.bookingType }} with {{ target.name }}`. 
BOOKING_BOOKED: `{{ actor.name }} booked {{ fields.bookingType }} with {{ target.name }}`. 
PROFILE_MODIFIED: `{{ actor.name }} changed profile picture`. 

etc..

I'm just waiting on the template feature to be fixed/released by the BoxyHQ team. Ive already contacted them.

Other than that, I think the core structure and UX/UI has been laid out. 

btw i recorded [this video](https://www.youtube.com/watch?v=Nbi8Tr1Ny9E) that show UI/UX and dx when developing integrations

cc: @zomars

oliverqx committed:
chore: update types

oliverqx committed:
feat: implement BOOKING_CONFIRMED and BOOKING_PAID

oliverqx committed:
feat: implement BOOKING_PAYMENT_INITIATED and BOOKING_REQUESTED

oliverqx committed:
fix: remove comments

oliverqx committed:
fix: implement BOOKING_RESCHEDULED

oliverqx committed:
fix: add missing user id

oliverqx committed:
feat: add General as part of DefaultAppSettingsOption

This panel will contain global settings options like factory reset.

oliverqx committed:
feat: add DefaultAppSettingsOptions.General related constants

oliverqx committed:
feat: Add DefaultAppSettingsOptions.GeneralSettings related utilities

oliverqx committed:
fix(ui): remove sticky from navigation panel

oliverqx committed:
feat: create general settings component

oliverqx committed:
fix: add -1 as id when no userId is present

oliverqx committed:
fix: remove unused type

oliverqx committed:
feat: render general settings

oliverqx committed:
fix: improve types

- Preparing to create audit log app context

oliverqx committed:
chore: update zod schemas

oliverqx committed:
feat: create and implement context

oliverqx committed:
refactor: implement context

oliverqx committed:
chore: improve dx at AuditLogEventToggles

oliverqx committed:
feat: add loading state to AuditSystemStatus

oliverqx committed:
feat: add setup step to app template

oliverqx committed:
fix: update default event and fix typo in logger

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
feat: add system events to schema

oliverqx committed:
feat: trpc audit logger v0.1

oliverqx committed:
chore: remove settings column from Credential

- settings and keys go in appKey

oliverqx committed:
feat: modify and create schema

- modify: remove AuditLogSystemTriggerEvents.SYSTEM_TEMPLATE_UPDATED
- modify: add AuditLogSystemTriggerEvents.SYSTEM_MISC to allow for implementation specific events without modifying schema
- add: AuditLogAppTriggerEvents.APP_KEYS_UPDATED

oliverqx committed:
chore: remove credentialSettingsSchema

oliverqx committed:
fix: add disabledEvents to genericAuditLogClient

oliverqx committed:
chore: add type

oliverqx committed:
fix: add credential id to auditLogManager and improve dx

oliverqx committed:
feat: improve handleAuditLogTrigger base design

oliverqx committed:
feat: add pathToAuditLogEvent

based on path this will return the events metadata, should be moved to core later

oliverqx committed:
feat: add getAuditLogManager test

oliverqx committed:
feat: add handleAuditLogTrigger test

oliverqx committed:
feat: add flattenObject

- avoided recursion so its a loop by choice

oliverqx committed:
feat: update handleAuditLogTrigger on handleCancelBooking

oliverqx committed:
feat: add buildCredential to test utils

oliverqx committed:
feat: update handleAuditLogTrigger on trpc plugin

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
feat: create AuditLogManager test

oliverqx committed:
fix: add string to array zod type

oliverqx committed:
feat: add system events interceptor to audit log feature core

oliverqx committed:
feat: trpc plugin strip return value of sensitive data

oliverqx committed:
refactor: move trpc test utils to their own file

oliverqx committed:
feat: add parser to credential

oliverqx committed:
feat: add handleAuditLogTrigger tests

- SYSTEM_CREDENTIALS_UPDATED
- SYSTEM_EVENT_OFF
- SYSTEM_EVENT_ON
- UPDATED_APP_KEYS

oliverqx committed:
fix: remove unused trpc endpoint

Credential.settings removed from schema

oliverqx committed:
feat: add App and Credential trigger events

- separated app actions from credentials

oliverqx committed:
feat: add default prisma and return data to be used by auditLogTrigger

oliverqx committed:
feat: create buildApp function in test utils

oliverqx committed:
feat: add saveKeys and toggleApp + change updateAppCredentials

oliverqx committed:
feat: add logger and CREDENTIAL as target case to createEvent

oliverqx committed:
feat: create saveKeys and toggle tests + update previous test to now separated APP | CREDENTIAL

oliverqx committed:
refactor: improve dx

oliverqx committed:
feat: add AuditLogApiKeysTriggerEvents

oliverqx committed:
feat: add logger

oliverqx committed:
refactor: dry and add case for API_KEYS

oliverqx committed:
fix: remove boxyHQAuditLogManager

oliverqx committed:
fix: change action -> trigger for better naming consistency across feature

oliverqx committed:
feat: add triggerToMetadata for all API_KEYS triggers

oliverqx committed:
feat: add user to return value, will be used by handleAuditLogTrigger

oliverqx committed:
feat: add relevant data to all trpc endpoints related to API_KEYS triggers

oliverqx committed:
feat: call handleAuditLogTrigger when API_KEY is used

oliverqx committed:
fix: make array contents optional as array can be empty

oliverqx committed:
feat: create apiKeyEventsReporting test

oliverqx committed:
refactor: change test file name

oliverqx committed:
feat: add triggerMetadata to all BOOKING events

oliverqx committed:
feat: create unit testing for all BOOKING events

oliverqx committed:
feat: implement auditLogTrigger BOOKING events

oliverqx committed:
fix: update imports

oliverqx committed:
feat: add trpc middleware

oliverqx committed:
feat: update app handlers return value

oliverqx committed:
feat: createCallerFactory, will be used for integration testing

oliverqx committed:
fix: add default role to mock user

oliverqx committed:
feat: add buildSession test helper

oliverqx committed:
refactor: move getAuditLogManager unit test

oliverqx committed:
feat: update apiKeys trpc handlers return value to include trigger

oliverqx committed:
fix: update constant key values from path to enum

oliverqx committed:
refactor: move auditlog booking unit tests

oliverqx committed:
refactor: API_KEY trpc handler integration tests

oliverqx committed:
refactor: SYSTEM and CREDENTIAL trpc integration tests

oliverqx committed:
feat: add log.silly

oliverqx committed:
fix: trpc-plugin manage auditLog data properly

oliverqx committed:
feat: add AuditLogWebhookTriggerEvents

oliverqx committed:
feat: add webhook triggers to audit-log feature utilities

oliverqx committed:
feat: prepare webhook trpc router for audit logging

oliverqx committed:
feat: create webhook trpc <-> auditLog integration testing

oliverqx committed:
fix: add types

oliverqx committed:
fix: add types

oliverqx committed:
fix: types

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx commented:
TODO: 

- [ ] Documentation
- [ ] Missing events
- [ ] Typing
- [ ] Synchronize with old UI

Missing Events
- [ ]   SCHEDULE
- [ ]   EVENT_TYPES
- [ ]   TEAMS
- [ ]   USER_PROFILE
- [ ]   PASSWORD
- [ ]   IMPERSONATION
- [ ]   ROUTING_FORMS
- [ ]   WORKFLOWS
- [ ] - [x]   BOOKING
- [x]   WEBHOOKS
- [x]   SYSTEM
- [x]   API_KEYS - MISSING IS API_KEY_USED
- [x]   APPS
- [ ]   CREDENTIAL  - (installed / deleted missing)

oliverqx committed:
chore: improve types

oliverqx committed:
Merge branch 'feature.logging' of https://github.com/oliverqx/cal.com into feature.logging

oliverqx committed:
fix: typesafety trpc-plugin

oliverqx committed:
fix: update apikeys handlers

oliverqx committed:
fix: remove unused code

oliverqx committed:
fix: types

oliverqx committed:
fix: update return data to comply with shape expected by trpc-plugin

oliverqx committed:
fix: types

oliverqx committed:
fix: types

oliverqx committed:
feat: API_KEY_USED integration test

oliverqx committed:
chore: improve dx

oliverqx committed:
fix: update handleAuditLogTrigger

oliverqx committed:
fix: remove manual auditlogtrigger and use trpc-plugin

oliverqx committed:
fix: migration typo

oliverqx committed:
fix: ZAuditLogEventGroup.id is string

oliverqx committed:
fix: import and fix mock

oliverqx committed:
fix: move buildProfileMockData to builders

oliverqx committed:
fix: update auditLogManager naming AuditLogManager.test

oliverqx committed:
fix: remove unused arguments and add custom interceptor

oliverqx committed:
fix: uncomment code, remove console.log

oliverqx committed:
fix: typo in migration file

oliverqx committed:
feat: update appCredentialById

oliverqx committed:
feat: clean up audit-log-implementation UI

oliverqx committed:
fix: type issues

oliverqx committed:
fix: remove console.log

oliverqx committed:
fix: add error and add type assertion

oliverqx committed:
chore: add type

oliverqx committed:
fix: add type assertion

oliverqx committed:
Merge branch 'main' into feature.logging

oliverqx committed:
fix: update import

oliverqx committed:
Merge branch 'feature.logging' of https://github.com/oliverqx/cal.com into feature.logging

oliverqx commented:
Closed PR as I've been working on a library of my own that I will use to solve this bounty. Opening PR soon
